pkg/providers/hostresourcegroup/hostresourcegroup.go

```diff
@@ -0,0 +1,94 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hostresourcegroup
+
+import (
+	"context"
+	"sync"
+
+	//"github.com/aws/aws-sdk-go/service/resourcegroups"
+	"github.com/aws/aws-sdk-go/service/resourcegroups"
+	"github.com/aws/aws-sdk-go/service/resourcegroups/resourcegroupsiface"
+	"github.com/aws/karpenter-core/pkg/utils/pretty"
+	"github.com/aws/karpenter/pkg/apis/v1beta1"
+	"github.com/patrickmn/go-cache"
+	"knative.dev/pkg/logging"
+)
+
+type Provider struct {
+	sync.RWMutex
+	resourcegroups resourcegroupsiface.ResourceGroupsAPI
+	cache          *cache.Cache
+	cm             *pretty.ChangeMonitor
+}
+
+func NewProvider(rgapi resourcegroupsiface.ResourceGroupsAPI, cache *cache.Cache) *Provider {
+	return &Provider{
+		resourcegroups: rgapi,
+		cm:             pretty.NewChangeMonitor(),
+		// TODO: Remove cache for v1beta1, utilize resolved subnet from the AWSNodeTemplate.status
+		// Subnets are sorted on AvailableIpAddressCount, descending order
+		cache: cache,
+	}
+}
+
+func (p *Provider) Get(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) (*v1beta1.HostResourceGroup, error) {
+	p.Lock()
+	defer p.Unlock()
+    var nextToken *string = nil
+	var groups []*resourcegroups.GroupIdentifier
+	for {
+		resp, err := p.resourcegroups.ListGroupsWithContext(ctx, &resourcegroups.ListGroupsInput{ NextToken: nextToken})
+		if err != nil {
+            logging.FromContext(ctx).
+                With("aws error", err).
+                Debugf("Error from resourcegroup:listgroups")
+			return nil, err
+		}
+        nextToken = resp.NextToken
+        for i := range resp.GroupIdentifiers {
+            groups = append(groups, resp.GroupIdentifiers[i])
+        }
+		if nextToken == nil {
+			break
+		}
+	}
+
+	// filter resp to only include those that match the hostResourceGroupSelector Name
+	for i := range groups {
```

Comment:
Same comment on using pages for this rather than building out everything in memory and then iterating through it later. That'll help memory performance